pkg/bgpv1/manager/reconcilerv2/pod_cidr.go

```diff
@@ -0,0 +1,196 @@
+	// get desired paths per address family
+	desiredFamilyAdverts, err := r.getDesiredPathsPerFamily(p, podCIDRPrefixes)
+	if err != nil {
+		return err
+	}
+
+	return r.reconcileFamilyAdvertisements(ctx, p, desiredFamilyAdverts)
+}
+
+// getDesiredPathsPerFamily returns a map of desired paths per address family.
+// Note: This returns prefixes per address family. Global routing table will contain prefix per family not per neighbor.
+// Per neighbor advertisement will be controlled by BGP Policy.
+func (r *PodCIDRReconciler) getDesiredPathsPerFamily(p ReconcileParams, desiredPrefixes []netip.Prefix) (PodCIDRMetadata, error) {
+	// get per peer per family pod cidr advertisements
+	desiredPeerAdverts, err := r.peerAdvert.GetConfiguredAdvertisements(p.DesiredConfig, v2alpha1.BGPPodCIDRAdvert)
+	if err != nil {
+		return nil, err
+	}
+
+	// Calculate desired paths per address family, collapsing per-peer advertisements into per-family advertisements.
+	desiredFamilyAdverts := make(PodCIDRMetadata)
+	for _, peerFamilyAdverts := range desiredPeerAdverts {
+		for family, familyAdverts := range peerFamilyAdverts {
+			if len(familyAdverts) == 0 {
+				// if no POD CIDR advertisement is present for a peer,
+				// we can skip the peer and continue with the next peer.
+				// We do not need to evaluate individual advertisement.
+				continue
+			}
+
+			agentFamily := types.ToAgentFamily(family)
+
+			// create existingPathsSet for easier existence check.
+			existingPathsSet := sets.New[string]()
+			existingPaths, ok := desiredFamilyAdverts[agentFamily]
+			if ok {
+				for _, existingPath := range existingPaths {
+					existingPathsSet.Insert(existingPath.NLRI.String())
+				}
+			}
+
+			var paths []*types.Path
+			for _, prefix := range desiredPrefixes {
+				path := types.NewPathForPrefix(prefix)
+				path.Family = agentFamily
+
+				// skip if path already exists.
+				if existingPathsSet.Has(path.NLRI.String()) {
+					continue
+				}
+				paths = append(paths, path)
+			}
+
+			desiredFamilyAdverts[types.ToAgentFamily(family)] = append(desiredFamilyAdverts[types.ToAgentFamily(family)], paths...)
+		}
+	}
+	return desiredFamilyAdverts, nil
+}
+
+func (r *PodCIDRReconciler) reconcileFamilyAdvertisements(ctx context.Context, p ReconcileParams, desiredFamilyAdverts PodCIDRMetadata) error {
+	l := r.logger.WithFields(logrus.Fields{
+		types.ReconcilerLogField: r.Name(),
+		types.InstanceLogField:   p.DesiredConfig.Name,
+	})
+
+	runningFamilyAdverts := r.getMetadata(p.BGPInstance)
+
+	// to delete family advertisements that are not in desiredFamilyAdverts
+	for family := range runningFamilyAdverts {
+		if _, ok := desiredFamilyAdverts[family]; !ok {
+			_, err := ReconcileAdvertisement(&AdvertisementsReconcilerParams{
+				Logger:                l,
+				Ctx:                   ctx,
+				Instance:              p.BGPInstance,
+				CurrentAdvertisements: runningFamilyAdverts[family],
+				ToAdvertise:           nil,
+			})
+			if err != nil {
+				return err
```

Comment:
Modified Advertisement reconciler. It will return running advertisements in cases where there is an error.

I don't think we have a good way to test this. I will think more about it and add a test case in follow up PR.